thrall/app/lib/ThrallStreamProcessor.scala

```diff
@@ -1,62 +1,101 @@
+    case withMarker:LogMarker => withMarker.markerContents
+    case _ => Map()
+  }) ++ Map(
     "recordPriority" -> priority.toString,
-    "recordArrivalTime" -> DateTimeUtils.toString(record.approximateArrivalTimestamp)
+    "recordArrivalTime" -> DateTimeUtils.toString (arrivalTimestamp)
   )
+
+  def map[V](f: T => V): TaggedRecord[V] = this.copy(payload = f(payload))
 }
 
 class ThrallStreamProcessor(
-  highPrioritySource: Source[KinesisRecord, Future[Done]],
-  lowPrioritySource: Source[KinesisRecord, Future[Done]],
+  uiSource: Source[KinesisRecord, Future[Done]],
+  automationSource: Source[KinesisRecord, Future[Done]],
+  reingestionSource: Source[ReingestionRecord, Future[Done]],
   consumer: ThrallEventConsumer,
   actorSystem: ActorSystem,
-  materializer: Materializer) extends GridLogging {
+  materializer: Materializer
+ ) extends GridLogging {
 
   implicit val mat = materializer
   implicit val dispatcher = actorSystem.getDispatcher
 
-  val mergedKinesisSource: Source[TaggedRecord, NotUsed] = Source.fromGraph(GraphDSL.create() { implicit g =>
+  val mergedKinesisSource: Source[TaggedRecord[UpdateMessage], NotUsed] = Source.fromGraph(GraphDSL.create() { implicit graphBuilder =>
     import GraphDSL.Implicits._
-    val highPriorityKinesisSource = highPrioritySource.map(TaggedRecord(_, HighPriority))
-    val lowPriorityKinesisSource = lowPrioritySource.map(TaggedRecord(_, LowPriority))
 
-    val mergePreferred = g.add(MergePreferred[TaggedRecord](1))
+    val uiRecordSource = uiSource.map(kinesisRecord =>
+      TaggedRecord(kinesisRecord.data.toArray, kinesisRecord.approximateArrivalTimestamp, UiPriority, kinesisRecord.markProcessed))
+
+    val automationRecordSource = automationSource.map(kinesisRecord =>
+      TaggedRecord(kinesisRecord.data.toArray, kinesisRecord.approximateArrivalTimestamp, AutomationPriority, kinesisRecord.markProcessed))
 
-    highPriorityKinesisSource ~> mergePreferred.preferred
-    lowPriorityKinesisSource  ~> mergePreferred.in(0)
+    val reingestionMessagesSource: Source[TaggedRecord[UpdateMessage], Future[Done]] = reingestionSource.map { case ReingestionRecord(updateMessage, time) =>
+      TaggedRecord(updateMessage, time, ReingestionPriority, () => {})
+    }
+
+    // merge together ui and automation kinesis records
+    val uiAndAutomationRecordsMerge = graphBuilder.add(MergePreferred[TaggedRecord[Array[Byte]]](1))
+    uiRecordSource ~> uiAndAutomationRecordsMerge.preferred
+    automationRecordSource  ~> uiAndAutomationRecordsMerge.in(0)
+
+    // parse the kinesis records into thrall update messages (dropping those that fail)
+    val uiAndAutomationMessagesSource: PortOps[TaggedRecord[UpdateMessage]] =
+      uiAndAutomationRecordsMerge.out
+        .map { taggedRecord =>
+          ThrallEventConsumer
+            .parseRecord(taggedRecord.payload, taggedRecord.arrivalTimestamp)
+            .map(
+              message => taggedRecord.copy(payload = message)
```

Comment:
Looks like `payload` can represent both a raw message and its parsed version. Would it make sense to model the raw and parsed versions separately, to ensure they're never confused by consuming code?

Reply:
I see where you're coming from, however as the comment in https://github.com/guardian/grid/pull/3271/commits/b44df43a3f8aa4cdf335e86c2c7b8ceff330f431 explains, `TaggedRecord` is used to represent both Kinesis records and messages originating in thrall which wouldn't have a raw payload (so we'd have to model those as Option which would be messy). The `P` type param of `TaggedRecord` is sufficient for consumers to know whether it's been parsed or not (`Array[Byte]` when raw, and `UpdateMessage` when parsed, compiler will tell the user when it's the wrong shape.

Reply:
Ah yep, good point re: the type param 👍